### PR TITLE
[SPARK-45130][CONNECT][ML][PYTHON] Avoid Spark connect ML model to change input pandas dataframe

### PR DESCRIPTION
**Pull request by WeichenXu123**
### What changes were proposed in this pull request?

Currently, to avoid data copy, Spark connect ML model directly changes input pandas dataframe for appending prediction columns. But we can use `pandas_df.copy(deep=False)` to shallow copy it and then append prediction columns in copied dataframe. This is easier for user to use it.

### Why are the changes needed?

This makes `pyspark.ml.connect` model `transform` method has more similar behavior with `pyspark.ml` model, i.e., the input dataframe is intact after `transform` is called. Otherwise user might be surprise at the new behavior and have to change more code to migrate their workload to `pyspark.ml.connect`

### Does this PR introduce _any_ user-facing change?

Yes. 
Previous behavior:
In `pyspark.ml.connect`, `model.transform` will append new columns into input pandas dataframe, and return input dataframe object.

Changed behavior:
In `pyspark.ml.connect`, `model.transform` will shallow copy input pandas dataframe and append new columns into shallow copied pandas dataframe, then return copied pandas dataframe.

### How was this patch tested?

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/ml/tests/connect/test_legacy_mode_feature.py [hunk 1]
```diff
@@ -57,7 +57,10 @@ def test_max_abs_scaler(self):
 
         local_df1 = df1.toPandas()
         local_fit_model = scaler.fit(local_df1)
+        input_cols = local_df1.columns.tolist()
         local_transform_result = local_fit_model.transform(local_df1)
+        assert local_df1.columns.tolist() == input_cols, \
+            "pandas dataframe input columns should be intact."
```
Comment (harupy) [inline python/pyspark/ml/tests/connect/test_legacy_mode_feature.py:63]:
```suggestion
        input_cols = local_df1.columns.copy()
        local_transform_result = local_fit_model.transform(local_df1)
        assert local_df1.columns.equals(input_cols), \
            "pandas dataframe input columns should be intact."
```

Comment (harupy) [reply python/pyspark/ml/tests/connect/test_legacy_mode_feature.py:63]: Can we compare dataframes to ensure nothing has changed?

**Comment by HyukjinKwon** [comment]
https://github.com/WeichenXu123/spark/runs/16710260795

File: python/pyspark/ml/tests/connect/test_legacy_mode_feature.py [hunk 2]
```diff
@@ -57,8 +58,10 @@ def test_max_abs_scaler(self):
 
         local_df1 = df1.toPandas()
         local_fit_model = scaler.fit(local_df1)
+        local_df1_copy = local_df1.copy()
         local_transform_result = local_fit_model.transform(local_df1)
-        assert id(local_transform_result) == id(local_df1)
+        # assert that pandas dataframe input columns should be intact.
+        pd.testing.assert_frame_equal(local_df1, local_df1_copy, check_type=False)
```
Comment (harupy) [inline python/pyspark/ml/tests/connect/test_legacy_mode_feature.py:64]: why do we need `check_type`?

File: python/pyspark/ml/tests/connect/test_legacy_mode_feature.py [hunk 3]
```diff
@@ -57,8 +58,10 @@ def test_max_abs_scaler(self):
 
         local_df1 = df1.toPandas()
         local_fit_model = scaler.fit(local_df1)
+        local_df1_copy = local_df1.copy()
         local_transform_result = local_fit_model.transform(local_df1)
-        assert id(local_transform_result) == id(local_df1)
+        # assert that pandas dataframe input columns should be intact.
```
Comment (harupy) [inline python/pyspark/ml/tests/connect/test_legacy_mode_feature.py:63]:
```suggestion
        # assert that `transform` doesn't mutate the input dataframe.
```

File: python/pyspark/ml/connect/base.py [hunk 4]
```diff
@@ -154,8 +154,9 @@ def transform(
         The dataset can be either pandas dataframe or spark dataframe,
         if it is a spark DataFrame, the result of transformation is a new spark DataFrame
         that contains all existing columns and output columns with names.
-        if it is a pandas DataFrame, the input pandas dataframe is appended with output
-        columns in place.
+        if it is a pandas DataFrame, the input pandas dataframe is intact, and the output
+        dataframe shallow copies all existing columns from input dataframe and appends
+        output columns.
```
Comment (harupy) [inline python/pyspark/ml/connect/base.py:159]:
```suggestion
        If it is a pandas DataFrame, the result of transformation is a shallow copy
        of the input pandas dataframe with output columns with names.
```

to be consistent?

**Review by harupy — APPROVED** [review]
LGTM!

File: python/pyspark/ml/tests/connect/test_legacy_mode_classification.py [hunk 5]
```diff
@@ -19,6 +19,7 @@
 import tempfile
 import unittest
 import numpy as np
+import pandas as pd
```
Comment (HyukjinKwon) [inline python/pyspark/ml/tests/connect/test_legacy_mode_classification.py:22]: can you import this within `test_binary_classes_logistic_regression` or under `should_test_connect` so the tests can be skipped fine when `pandas` is not installed?

File: python/pyspark/ml/tests/connect/test_legacy_mode_feature.py [hunk 6]
```diff
@@ -19,6 +19,7 @@
 import os
 import pickle
 import numpy as np
+import pandas as pd
```
Comment (HyukjinKwon) [inline python/pyspark/ml/tests/connect/test_legacy_mode_feature.py:22]: ditto

File: python/pyspark/ml/tests/connect/test_legacy_mode_pipeline.py [hunk 7]
```diff
@@ -19,6 +19,7 @@
 import tempfile
 import unittest
 import numpy as np
+import pandas as pd
```
Comment (HyukjinKwon) [inline python/pyspark/ml/tests/connect/test_legacy_mode_pipeline.py:22]: ditto

File: python/pyspark/ml/tests/connect/test_legacy_mode_classification.py (hunk 5 above)
Comment (WeichenXu123) [reply python/pyspark/ml/tests/connect/test_legacy_mode_classification.py:22]: isn't pandas always installed in CI image ?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged to master

**Pull request closed without merging**